### PR TITLE
Develop/features/dlsv2 437 add result review request links to supervisor to do list

### PR DESCRIPTION
**Pull request by daniel-manta-hee**
### JIRA link
https://hee-dls.atlassian.net/browse/DLSV2-437

### Description
Added new query in Supervisor service for retrieving requested reviews which would work in the same way as the existing signoff requested to do list. Updated the controller method and the view. 

### Screenshots
image

-----
### Developer checks
Checked that this code returns one record for competency with pending review.
```sql
DECLARE @adminId INT = 1
SELECT sa.id AS SelfAssessmentId, co.ID as CompetencyId, sar.ID AS RecordId, ca.ID, sd.ID AS SupervisorDelegateId, c.FirstName + ' ' + c.LastName AS DelegateName, sa.Name AS ProfileName, sasv.Requested, 0 AS SignOffRequest, 1 AS ResultsReviewRequest
    FROM   CandidateAssessmentSupervisors AS cas INNER JOIN
    CandidateAssessments AS ca ON cas.CandidateAssessmentID = ca.ID INNER JOIN
    Candidates AS c ON ca.CandidateID = c.CandidateID INNER JOIN
    SelfAssessments AS sa ON ca.SelfAssessmentID = sa.ID INNER JOIN
    SupervisorDelegates AS sd ON cas.SupervisorDelegateId = sd.ID INNER JOIN
    SelfAssessmentResults AS sar ON sar.SelfAssessmentID = sa.ID INNER JOIN
    Competencies AS co ON sar.CompetencyID = co.ID INNER JOIN					
    SelfAssessmentResultSupervisorVerifications AS sasv ON sasv.SelfAssessmentResultId = sar.ID AND sasv.CandidateAssessmentSupervisorID = cas.ID AND sar.DateTime = (
		SELECT MAX(sar2.DateTime)
		FROM SelfAssessmentResults AS sar2
		WHERE sar2.SelfAssessmentID = sar.SelfAssessmentID AND sar2.CompetencyID = co.ID
	)
WHERE (sd.SupervisorAdminID = @adminId) AND (sasv.Verified IS NULL)
```
image

Grouped by SelfAssessment in order to get only one record per self assessment.
```sql
DECLARE @adminId INT = 1
SELECT ca.ID, sd.ID AS SupervisorDelegateId, c.FirstName + ' ' + c.LastName AS DelegateName, sa.Name AS ProfileName, MAX(sasv.Requested) AS Requested, 0 AS SignOffRequest, 1 AS ResultsReviewRequest
      FROM   CandidateAssessmentSupervisors AS cas INNER JOIN
      CandidateAssessments AS ca ON cas.CandidateAssessmentID = ca.ID INNER JOIN
      Candidates AS c ON ca.CandidateID = c.CandidateID INNER JOIN
      SelfAssessments AS sa ON ca.SelfAssessmentID = sa.ID INNER JOIN
      SupervisorDelegates AS sd ON cas.SupervisorDelegateId = sd.ID INNER JOIN
SelfAssessmentResults AS sar ON sar.SelfAssessmentID = sa.ID INNER JOIN
Competencies AS co ON sar.CompetencyID = co.ID INNER JOIN					
      SelfAssessmentResultSupervisorVerifications AS sasv ON sasv.SelfAssessmentResultId = sar.ID AND sasv.CandidateAssessmentSupervisorID = cas.ID AND sar.DateTime = (
SELECT MAX(sar2.DateTime)
FROM SelfAssessmentResults AS sar2
WHERE sar2.SelfAssessmentID = sar.SelfAssessmentID AND sar2.CompetencyID = co.ID
)
  WHERE (sd.SupervisorAdminID = @adminId) AND (sasv.Verified IS NULL)
GROUP BY sa.ID, ca.ID, sd.ID, c.FirstName, c.LastName, sa.Name
```
image

### CODE REVIEW

File: DigitalLearningSolutions.Web/Views/Supervisor/Index.cshtml [hunk 1]
```diff
@@ -67,17 +67,28 @@
...
           Review and sign-off @toDoItem.DelegateName's @toDoItem.ProfileName
         </a>. Requested @toDoItem.Requested.ToShortDateString()
-
       </li>
     }
-  </ul>
+    else if(toDoItem.ResultsReviewRequest)
+    {
+      <li class="nhsuk-contents-list__item">
+        <a class="nhsuk-contents-list__link nhsuk-link--no-visited-state" asp-action="VerifyDelegateSelfAssessment" asp-route-supervisorDelegateId="@toDoItem.SupervisorDelegateId" asp-route-candidateAssessmentId="@toDoItem.ID">
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Web/Views/Supervisor/Index.cshtml:85]: Could we use the same action and routing as the "Review and sign-off" link above, here. The link is broken for me with the current action and route data and should go to the same place.

File: DigitalLearningSolutions.Web/Views/Supervisor/Index.cshtml [hunk 2]
```diff
@@ -67,17 +67,28 @@
...
+          Verify self-assessment results for @toDoItem.DelegateName's @toDoItem.ProfileName
+        </a>. Requested @toDoItem.Requested.ToShortDateString()
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Web/Views/Supervisor/Index.cshtml:87]: The ". Requested @toDoItem.Requested..." text falls onto a new blank line if the '<a>' text is long:
image
Could we fix this to keep it in-line using CSS or move the Requested date inside the link, if not?

**Review by kevwhitt-hee — CHANGES_REQUESTED** [review]
A couple of changes requested to the Index.cshtml - see comments.

File: DigitalLearningSolutions.Web/Views/Supervisor/Index.cshtml (hunk 2 above)
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Web/Views/Supervisor/Index.cshtml:87]: We should do this for the other type of ToDo item, too.

**Comment by daniel-manta-hee** [comment]
Yes, text break comes from class nhsuk-contents-list__link on the link. I believe we could just remove that class from the html tag. It only seems to be doing the "display: inline-block;".
________________________________

**Comment by kevwhitt-hee** [comment]
Yes. Let's remove the "nhsuk-contents-list__link" from both links, please. I've tested this in browser dev mode and it looks much better,

File: DigitalLearningSolutions.Data/Services/SupervisorService.cs [hunk 3]
```diff
@@ -320,6 +321,26 @@ public IEnumerable<SupervisorDashboardToDoItem> GetSupervisorDashboardToDoItems(
                 WHERE (sd.SupervisorAdminID = @adminId) AND (casv.Verified IS NULL)", new { adminId }
                 );
         }
+        public IEnumerable<SupervisorDashboardToDoItem> GetSupervisorDashboardToDoItemsForRequestedReviews(int adminId)
+        {
+            return connection.Query<SupervisorDashboardToDoItem>(
+                @"SELECT ca.ID, sd.ID AS SupervisorDelegateId, c.FirstName + ' ' + c.LastName AS DelegateName, sa.Name AS ProfileName, MAX(sasv.Requested) AS Requested, 0 AS SignOffRequest, 1 AS ResultsReviewRequest
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Data/Services/SupervisorService.cs:327]: Further testing suggests there is a problem with this query. I requested verification for 1 result in a second self assessment. This appeared on my supervisor To Do list three times:
image

Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Data/Services/SupervisorService.cs:327]: The return from this query has the following return on my Dev DB for adminId = 1:
image
Note the different IDs (CandidateAssessment.IDs?) in the first column. Suggesting a problem with a join.

**Review by kevwhitt-hee — APPROVED** [review]
Working perfectly now.

**Pull request merged**